### PR TITLE
Protect post relayed call

### PR DESCRIPTION
**Pull request by shahafn**
See issues #54 #94

### CODE REVIEW

File: contracts/RelayHub.sol [hunk 1]
```diff
@@ -152,16 +154,27 @@ contract RelayHub is RelayHubApi {
...
+        bytes memory packed = abi.encodePacked("rlx:", from, to, encoded_function, transaction_fee, gas_price, gas_limit, nonce, address(this));
         bytes32 hashed_message = keccak256(abi.encodePacked(packed, relay));
         bytes32 signed_message = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", hashed_message));
         if (!GsnUtils.checkSig(from, signed_message,  approval))  // Verify the sender's signature on the transaction
             return 1;   // @from hasn't signed the transaction properly
         if (nonces[from] != nonce)
             return 2;   // Not a current transaction.  May be a replay attempt.
         // XXX check @to's balance, roughly estimate if it has enough balance to pay the transaction fee.  It's the relay's responsibility to verify, but check here too.
-        return to.accept_relayed_call(relay, from, transaction, gas_price, transaction_fee, approval); // Check to.accept_relayed_call, see if it agrees to accept the charges.
+        bytes memory transaction = abi.encodeWithSelector(to.accept_relayed_call.selector, relay, from, encoded_function, gas_price, transaction_fee, approval);
```
Comment (alex-forshtat-tbk) [inline contracts/RelayHub.sol]: 'transaction' is the most generic word in the Smart Contract code ever. Call it 't'. Or, 'a'. Or, don't.

File: contracts/RelayHub.sol [hunk 2]
```diff
@@ -181,29 +194,51 @@ contract RelayHub is RelayHubApi {
...
+        uint32 can_relay_result = can_relay(msg.sender, from, RelayRecipient(to), encoded_function, transaction_fee, gas_price, gas_limit, nonce, approval);
+        if (can_relay_result != 0) {
+            emit TransactionRelayed(msg.sender, from, to, keccak256(encoded_function), uint(RelayCallStatus.CanRelayFailed), 0);
+            return;
+        }
 
         // ensure that the last bytes of @transaction are the @from address.
         // Recipient will trust this reported sender when msg.sender is the known RelayHub.
-        bytes memory transaction = abi.encodePacked(encoded_function,from);
+//        bytes memory transaction = abi.encodePacked(encoded_function,from);
```
Comment (alex-forshtat-tbk) [inline contracts/RelayHub.sol]: cleanup comment line

File: contracts/RelayHub.sol [hunk 3]
```diff
@@ -181,29 +194,51 @@ contract RelayHub is RelayHubApi {
...
     }
 
-    function executeCallWithGas(uint allowed_gas, address to, uint256 value, bytes memory data) internal returns (bool success) {
-        assembly {
-            success := call(allowed_gas, to, value, add(data, 0x20), mload(data), 0, 0)
-        }
+    function recipient_calls(address from, address to, address relay_addr, bytes calldata encoded_function, uint transaction_fee, uint gas_limit, uint initial_gas) external returns (bool) {
+        require(msg.sender == address(this), "Only RelayHub should call this function");
+
+        // ensure that the last bytes of @transaction are the @from address.
```
Comment (alex-forshtat-tbk) [inline contracts/RelayHub.sol]: Move 2 lines down ;-)

File: contracts/RelayHub.sol (hunk 1 above)
Comment (shahafn) [reply contracts/RelayHub.sol]: Changed it to 'x'. 
We good?

File: contracts/RelayHub.sol (hunk 2 above)
Comment (shahafn) [reply contracts/RelayHub.sol]: done

File: contracts/RelayHub.sol (hunk 3 above)
Comment (shahafn) [reply contracts/RelayHub.sol]: done

**Pull request merged**